scheduler/rank.go

```diff
@@ -206,10 +206,34 @@ OUTER:
 			continue
 		}
 
-		// Index the existing network usage
+		// Index the existing network usage.
+		// This should never collide, since it represents the current state of
+		// the node. If it does collide though, it means we found a bug! So
+		// collect as much information as possible.
 		netIdx := structs.NewNetworkIndex()
-		netIdx.SetNode(option.Node)
-		netIdx.AddAllocs(proposed)
+		if collide, reason := netIdx.SetNode(option.Node); collide {
+			iter.sendEvent(&PortCollisionEvent{
+				Reason:   reason,
+				NetIndex: netIdx.Copy(),
+				Node:     option.Node,
+			})
+			iter.ctx.Metrics().ExhaustedNode(option.Node, "network: port collision")
+			continue
+		}
+		if collide, reason := netIdx.AddAllocs(proposed); collide {
+			event := &PortCollisionEvent{
+				Reason:      reason,
+				NetIndex:    netIdx.Copy(),
+				Node:        option.Node,
+				Allocations: make([]*structs.Allocation, len(proposed)),
+			}
+			for i, alloc := range proposed {
+				event.Allocations[i] = alloc.Copy()
```

Comment:
We need to either set `alloc.Job = nil` or sanitize `{Consul,Nomad,Vault}Token` fields out of the jobs.